src/cargo/ops/registry.rs

```diff
@@ -90,8 +92,37 @@ pub struct PublishOpts<'cfg> {
 
 pub fn publish(ws: &Workspace<'_>, opts: &PublishOpts<'_>) -> CargoResult<()> {
     let specs = opts.to_publish.to_package_id_specs(ws)?;
+
     let mut pkgs = ws.members_with_features(&specs, &opts.cli_features)?;
 
+    if let Packages(opt_in) = &opts.to_publish {
+        if ws.is_virtual() {
+            bail!("can't use \"--package <SPEC>\" in virtual manifest")
+        }
+        let matched_packages = ws
+            .members()
+            .filter(|m| specs.iter().any(|spec| spec.matches(m.package_id())))
+            .collect::<Vec<_>>();
+        if matched_packages.is_empty() {
+            bail!(
+                "not found `{}` in manifest members. Check in manifest path `{}`",
+                opt_in.get(0).unwrap(),
+                ws.root().display()
+            )
+        }
+        if matched_packages.len() > 1 {
+            bail!(
+                "found multiple `{}` in manifest members. Check in manifest path `{}`",
+                opt_in.get(0).unwrap(),
+                ws.root().display()
+            )
+        }
```

Comment:
There's a function [`PackageIdSpec::query`](https://github.com/rust-lang/cargo/blob/83a7983ec11c7239bdf790cf157c1dd7520f43fc/src/cargo/core/package_id_spec.rs#L188-L193) which already handles the "only one" logic. I think we should reuse that instead of implementing new logic. That will handle providing suggestions and such, and have a consistent error display. It will also handle the case with `default-members`, which could result in `Packages::Default` which this isn't handling.  Sorry if my previous suggestion was a little misleading.

That is, the logic should roughly be:

```rust
let member_ids = ws.members().map(|p| p.package_id());
let specs = opts.to_publish.to_package_id_specs(ws)?;
if specs.len() > 1 {
    bail!("the `-p` argument must be specified to select a single package to publish");
}
// Check that the spec matches exactly one member.
let _ = specs[0].query(member_ids)?;
```

I would also place an assert after the call to `members_with_features` that it only returns one member, just to double-check things.

Reply:
Thank you for explaining so much to me! I update something under your suggestion. Plz take a look when you free.